File: libraries/cm_push_notifications/src/main/java/com/tokopedia/notifications/factory/ActionNotification.java
package com.tokopedia.notifications.factory;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Build;
import android.support.v4.app.NotificationCompat;
import android.text.TextUtils;
import android.view.View;
import android.widget.RemoteViews;

import com.tokopedia.notifications.R;
import com.tokopedia.notifications.common.CMConstant;
import com.tokopedia.notifications.common.CMNotificationUtils;
import com.tokopedia.notifications.model.ActionButton;
import com.tokopedia.notifications.model.BaseNotificationModel;
import com.tokopedia.notifications.receiver.CMBroadcastReceiver;

import java.util.List;

/**
 * @author lalit.singh
 */
public class ActionNotification extends BaseNotification {

    ActionNotification(Context context, BaseNotificationModel baseNotificationModel) {
        super(context, baseNotificationModel);
    }

    @Override
    public Notification createNotification() {
        NotificationCompat.Builder builder = getNotificationBuilder();
        RemoteViews collapsedView = new RemoteViews(context.getApplicationContext().getPackageName()
                , R.layout.layout_collapsed);
        setCollapseData(collapsedView, baseNotificationModel, true);

        builder.setDeleteIntent(createDismissPendingIntent(baseNotificationModel.getNotificationId(), getRequestCode()));
        RemoteViews expandedView = new RemoteViews(context.getApplicationContext().getPackageName(),
                R.layout.layout_big_image);
        builder.setStyle(new NotificationCompat.DecoratedCustomViewStyle())
                .setCustomContentView(collapsedView)
                .setCustomBigContentView(expandedView);
        if (baseNotificationModel.getMedia() != null &&
                !TextUtils.isEmpty((baseNotificationModel.getMedia().getMediumQuality()))) {
            expandedView.setImageViewBitmap(R.id.img_big,
                    CMNotificationUtils.loadBitmapFromUrl(baseNotificationModel.getMedia().getMediumQuality()));
        }
        if (CMNotificationUtils.hasActionButton(baseNotificationModel)) {
            addActionButton(baseNotificationModel.getActionButton(), expandedView);
        }
        setCollapseData(expandedView, baseNotificationModel, false);
        return builder.build();
    }

    private void setCollapseData(RemoteViews remoteView, BaseNotificationModel baseNotificationModel, boolean isCollapsed) {

        if (isCollapsed) {
            if (TextUtils.isEmpty(baseNotificationModel.getIcon())) {
                Bitmap iconBitmap = getBitmap(baseNotificationModel.getIcon());
                if (null != iconBitmap) {
                    remoteView.setImageViewBitmap(R.id.iv_icon_collapsed, iconBitmap);
                } else {
                    remoteView.setImageViewBitmap(R.id.iv_icon_collapsed, getBitmapLargeIcon());
                }
            } else if (baseNotificationModel.getMedia() != null) {
                Bitmap iconBitmap = getBitmap(baseNotificationModel.getMedia().getMediumQuality());
                if (null != iconBitmap) {
                    remoteView.setImageViewBitmap(R.id.iv_icon_collapsed, iconBitmap);
                } else {
                    remoteView.setImageViewBitmap(R.id.iv_icon_collapsed, getBitmapLargeIcon());
                }
            } else {
                remoteView.setImageViewBitmap(R.id.iv_icon_collapsed, getBitmapLargeIcon());
            }
        }

        remoteView.setTextViewText(R.id.tv_collapse_title, CMNotificationUtils.getSpannedTextFromStr(baseNotificationModel.getTitle()));
        remoteView.setTextViewText(R.id.tv_collapsed_message, CMNotificationUtils.getSpannedTextFromStr(baseNotificationModel.getMessage()));
        remoteView.setOnClickPendingIntent(R.id.collapseMainView, createMainPendingIntent(baseNotificationModel,
                getRequestCode()));
    }

    private void addActionButton(List<ActionButton> actionButtonList, RemoteViews expandedView) {
        ActionButton actionButton;
        expandedView.setViewVisibility(R.id.ll_action, View.VISIBLE);
        for (int i = 0; i < actionButtonList.size(); i++) {
            actionButton = actionButtonList.get(i);
            switch (i) {
                case 0:
                    expandedView.setViewVisibility(R.id.tv_button1, View.VISIBLE);
                    expandedView.setTextViewText(R.id.tv_button1, actionButton.getText());
                    expandedView.setOnClickPendingIntent(R.id.tv_button1, getButtonPendingIntent(actionButton, getRequestCode()));
                    break;
                case 1:
                    expandedView.setViewVisibility(R.id.tv_button2, View.VISIBLE);
                    expandedView.setTextViewText(R.id.tv_button2, actionButton.getText());
                    expandedView.setOnClickPendingIntent(R.id.tv_button2, getButtonPendingIntent(actionButton, getRequestCode()));
                    break;
                case 2:
                    expandedView.setViewVisibility(R.id.tv_button3, View.VISIBLE);
                    expandedView.setTextViewText(R.id.tv_button3, actionButton.getText());
                    expandedView.setOnClickPendingIntent(R.id.tv_button3, getButtonPendingIntent(actionButton, getRequestCode()));
                    break;
            }
        }

    }

    private PendingIntent getButtonPendingIntent(ActionButton actionButton, int requestCode) {
        PendingIntent resultPendingIntent;
        Intent intent = new Intent(context, CMBroadcastReceiver.class);
        intent.setAction(CMConstant.ReceiverAction.ACTION_BUTTON);
        intent.putExtra(CMConstant.EXTRA_NOTIFICATION_ID, baseNotificationModel.getNotificationId());
        intent.putExtra(CMConstant.ReceiverExtraData.ACTION_BUTTON_APP_LINK, actionButton.getAppLink());
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            resultPendingIntent = PendingIntent.getBroadcast(
                    context,
                    requestCode,
                    intent,
                    PendingIntent.FLAG_UPDATE_CURRENT
            );
        } else {
            resultPendingIntent = PendingIntent.getBroadcast(
                    context,
                    requestCode,
                    intent,
                    PendingIntent.FLAG_UPDATE_CURRENT
            );
        }
        return resultPendingIntent;
    }
}

File: libraries/cm_push_notifications/src/main/java/com/tokopedia/notifications/factory/ActionNotification.kt
package com.tokopedia.notifications.factory

import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.graphics.Bitmap
import android.os.Build
import android.text.TextUtils
import android.view.View
import android.widget.RemoteViews
import androidx.core.app.NotificationCompat
import com.tokopedia.notifications.R
import com.tokopedia.notifications.common.CMConstant
import com.tokopedia.notifications.common.CMNotificationUtils
import com.tokopedia.notifications.model.ActionButton
import com.tokopedia.notifications.model.BaseNotificationModel

/**
 * @author lalit.singh
 */
internal class ActionNotification internal constructor(context: Context, baseNotificationModel: BaseNotificationModel)
    : BaseNotification(context, baseNotificationModel) {

    private val tvButtonResIds = arrayOf(R.id.tv_button1, R.id.tv_button2, R.id.tv_button3)
    private val llButtonResIds = arrayOf(R.id.ll_button1, R.id.ll_button2, R.id.ll_button3)
    private val ivButtonResIds = arrayOf(R.id.iv_button1, R.id.iv_button2, R.id.iv_button3)


    override fun createNotification(): Notification {
        val builder = notificationBuilder
        val collapsedView = getCollapsedView()
        setCollapseData(collapsedView, baseNotificationModel, true)
        var expandedView = getExpandedView()
        collapsedView.setOnClickPendingIntent(R.id.collapseMainView,
                createMainPendingIntent(baseNotificationModel, requestCode))
        builder.setStyle(NotificationCompat.DecoratedCustomViewStyle())
                .setCustomContentView(collapsedView)
                .setCustomBigContentView(expandedView)


        baseNotificationModel.media?.let {
            it.mediumQuality.isBlank().let { isBlank ->
                if (!isBlank) {
                    expandedView.setOnClickPendingIntent(R.id.img_big,
                            createMainPendingIntent(baseNotificationModel, requestCode))
                    expandedView.setViewVisibility(R.id.img_big, View.VISIBLE)
                    baseNotificationModel.media?.mediumQuality?.let { mq ->
                        if (mq.startsWith("http") || mq.startsWith("www"))
                            expandedView.setImageViewBitmap(R.id.img_big,
                                    CMNotificationUtils.loadBitmapFromUrl(baseNotificationModel.media?.mediumQuality))
                        else expandedView.setImageViewBitmap(R.id.img_big,
                                getBitmap(baseNotificationModel.media?.mediumQuality))
                    }
                    baseNotificationModel.actionButton.isNotEmpty().let { isNonEmpty ->
                        if (isNonEmpty)
                            expandedView.setViewVisibility(R.id.layout_collapsed, View.GONE)
                    }
                }
            }
        }

        if (CMNotificationUtils.hasActionButton(baseNotificationModel)) {
            addActionButton(baseNotificationModel.actionButton, expandedView)
        }
        setCollapseData(expandedView, baseNotificationModel, false)
        return builder.setDeleteIntent(createDismissPendingIntent(baseNotificationModel.notificationId, requestCode))
                .build()
    }

    private fun getExpandedView(): RemoteViews {
        var expandedView: RemoteViews
        expandedView = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
            RemoteViews(context.applicationContext.packageName,
                    R.layout.cm_layout_big_image)
        else RemoteViews(context.applicationContext.packageName,
                R.layout.cm_layout_big_image_pre_dark_mode)
        if (baseNotificationModel.media == null) {
            expandedView = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
                RemoteViews(context.applicationContext.packageName,
                        R.layout.cm_layout_action_button)
            else RemoteViews(context.applicationContext.packageName,
                    R.layout.cm_layout_action_button_pre_dark_mode)

        }
        return expandedView
    }

    private fun getCollapsedView(): RemoteViews {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
            RemoteViews(context.applicationContext.packageName, R.layout.cm_layout_collapsed)
        else
            RemoteViews(context.applicationContext.packageName, R.layout.cm_layout_collapsed_pre_dark_mode)
    }

    private fun setCollapseData(remoteView: RemoteViews, baseNotificationModel: BaseNotificationModel, isCollapsed: Boolean) {
        if (isCollapsed) {
            when {
                !TextUtils.isEmpty(baseNotificationModel.icon) -> {
                    val iconBitmap = getBitmap(baseNotificationModel.icon)
                    remoteView.setImageViewBitmap(R.id.iv_icon_collapsed, iconBitmap)
                }
                baseNotificationModel.media != null -> baseNotificationModel.media?.mediumQuality?.let { imageUrl ->
                    val iconBitmap = getBitmap(imageUrl)
                    remoteView.setImageViewBitmap(R.id.iv_icon_collapsed, iconBitmap)
                }
                else -> remoteView.setImageViewBitmap(R.id.iv_icon_collapsed, bitmapLargeIcon)
            }
        } else if (!TextUtils.isEmpty(baseNotificationModel.icon) && baseNotificationModel.media == null) {
            val iconBitmap = getBitmap(baseNotificationModel.icon)
            remoteView.setImageViewBitmap(R.id.iv_icon_collapsed, iconBitmap)
        }

        remoteView.setTextViewText(R.id.tv_collapse_title, CMNotificationUtils.getSpannedTextFromStr(baseNotificationModel.title))
        remoteView.setTextViewText(R.id.tv_collapsed_message, CMNotificationUtils.getSpannedTextFromStr(baseNotificationModel.message))
        remoteView.setOnClickPendingIntent(if (isCollapsed) R.id.collapseMainView else R.id.status_bar_latest_event_content, createMainPendingIntent(baseNotificationModel,
                requestCode))
        if (baseNotificationModel.media == null || baseNotificationModel.media?.mediumQuality == null
                || TextUtils.isEmpty(baseNotificationModel.media?.mediumQuality)) {
            remoteView.setViewVisibility(if (isCollapsed) R.id.tv_collapsed_message else R.id.tv_expanded_message, View.VISIBLE)
            remoteView.setViewVisibility(if (isCollapsed) R.id.tv_expanded_message else R.id.tv_collapsed_message, View.GONE)
            remoteView.setTextViewText(if (isCollapsed) R.id.tv_collapsed_message else R.id.tv_expanded_message,
                    CMNotificationUtils.getSpannedTextFromStr(if (isCollapsed) baseNotificationModel.message
                    else if (!TextUtils.isEmpty(baseNotificationModel.detailMessage)) baseNotificationModel.detailMessage
                    else baseNotificationModel.message))
        }
    }

    private fun addActionButton(actionButtonList: List<ActionButton>, expandedView: RemoteViews) {
        var actionButton: ActionButton
        expandedView.setViewVisibility(R.id.ll_action, View.VISIBLE)
        for (i in actionButtonList.indices) {
            actionButton = actionButtonList[i]
            expandedView.setViewVisibility(tvButtonResIds[i], View.VISIBLE)
            expandedView.setTextViewText(tvButtonResIds[i], actionButton.text)
            expandedView.setOnClickPendingIntent(llButtonResIds[i], getButtonPendingIntent(actionButton))
            if (!TextUtils.isEmpty(actionButton.actionButtonIcon)) {
                val bitmap: Bitmap? = getActionButtonBitmap(actionButton.actionButtonIcon!!)
                if (bitmap != null) {
                    expandedView.setViewVisibility(ivButtonResIds[i], View.VISIBLE)
                    expandedView.setImageViewBitmap(ivButtonResIds[i], bitmap)
                }
            }
        }
    }

    private fun getButtonPendingIntent(actionButton: ActionButton): PendingIntent {
        val intent = getBaseBroadcastIntent(context, baseNotificationModel)
        intent.action = CMConstant.ReceiverAction.ACTION_BUTTON
        intent.putExtra(CMConstant.ReceiverExtraData.ACTION_BUTTON_EXTRA, actionButton)
        return getPendingIntent(context, intent, requestCode)
    }


}
